File: packages/mcp-core/src/portal-links.ts
/**
 * Hosted-portal deep links minted from a customer session URL.
 *
 * Phase 1 suffixes the already-minted `customerUrl` in the SDK so the
 * MCP widget can land on the auto-recharge form without a new network
 * call or a platform deploy. The query params are the same ones the
 * customer-app manage page already parses (`tab=credits&intent=autorecharge`).
 *
 * Keep this constant in lockstep with
 * `apps/customer-app/src/pages/customer/manage/index.tsx`. A rename
 * there without a matching change here degrades to the Credits tab
 * without the form open — not a dead link, but not the intended
 * destination either.
 */

export const PORTAL_AUTO_RECHARGE_QUERY = 'tab=credits&intent=autorecharge'

/**
 * Suffix a minted customer-portal URL so it opens the auto-recharge
 * form (or the Credits summary when auto-recharge is already on).
 *
 * Returns `null` for a missing or non-http URL — the same guard
 * `narrate.ts` uses. No silent default: a missing portal URL yields
 * `null` and the widget renders status without an action.
 */
export function autoRechargeUrlFrom(customerUrl: string | null | undefined): string | null {
  if (typeof customerUrl !== 'string' || !/^https?:\/\//i.test(customerUrl)) {
    return null
  }
  const separator = customerUrl.includes('?') ? '&' : '?'
  return `${customerUrl}${separator}${PORTAL_AUTO_RECHARGE_QUERY}`
}

File: core/solvapay-core/src/mcp/portal_links.rs
//! Hosted-portal deep links minted from a customer session URL.
//!
//! Keep [`PORTAL_AUTO_RECHARGE_QUERY`] in lockstep with
//! `apps/customer-app/src/pages/customer/manage/index.tsx`.

/// Query that opens the auto-recharge form on the customer manage page.
pub const PORTAL_AUTO_RECHARGE_QUERY: &str = "tab=credits&intent=autorecharge";

/// Suffix a minted customer-portal URL so it opens the auto-recharge form.
///
/// Returns `None` for a missing or non-http URL — no silent default.
#[must_use]
pub fn auto_recharge_url_from(customer_url: Option<&str>) -> Option<String> {
    let url = customer_url?;
    let lower = url.to_ascii_lowercase();
    if !lower.starts_with("http://") && !lower.starts_with("https://") {
        return None;
    }
    let separator = if url.contains('?') { '&' } else { '?' };
    Some(format!("{url}{separator}{PORTAL_AUTO_RECHARGE_QUERY}"))
}

#[cfg(test)]
mod tests {
    #![allow(clippy::unwrap_used)]

    use super::auto_recharge_url_from;

    #[test]
    fn rejects_missing_and_non_http() {
        assert_eq!(auto_recharge_url_from(None), None);
        assert_eq!(auto_recharge_url_from(Some("")), None);
        assert_eq!(auto_recharge_url_from(Some("ui://portal")), None);
        assert_eq!(auto_recharge_url_from(Some("ftp://portal.example/s")), None);
    }

    #[test]
    fn suffixes_http_urls() {
        assert_eq!(
            auto_recharge_url_from(Some("https://portal.example/s")).as_deref(),
            Some("https://portal.example/s?tab=credits&intent=autorecharge")
        );
        assert_eq!(
            auto_recharge_url_from(Some("https://portal.example/s?foo=1")).as_deref(),
            Some("https://portal.example/s?foo=1&tab=credits&intent=autorecharge")
        );
        assert_eq!(
            auto_recharge_url_from(Some("HTTP://portal.example/s")).as_deref(),
            Some("HTTP://portal.example/s?tab=credits&intent=autorecharge")
        );
    }
}
